CDAP-5664 Use external dataset naming feature in hydrator-plugins for lineage/metadata etc

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-5664
Build : http://builds.cask.co/browse/HYP-DBT6

Created five classes : ReferencePluginConfig (which contains the referenceName property); ReferenceBatchSource, ReferenceBatchSink, ReferenceRealtimeSink, ReferenceRealtimeSource - these four classes validate whether the referenceName property satisfies the constraints for a dataset instance name. And in the case of realtime plugins, they also 'getDataset' to register lineage.

All the external source/sink plugins should extend one of the above four classes. Their PluginConfig classes should also extend ReferencePluginConfig class.

Unit Tests had to modified since referenceName is a non-nullable property and so it had to be included in the plugin properties.

Also, modified the widget JSONs to have the referenceName property.

gokulavasan committed:
CDAP-5664 Use external dataset naming feature in hydrator-plugins for lineage/metadata etc

Missed FileBatchSource

Make unit test work

Use referenceName while setting input/output

Remove space

Address PR comments

Addressing PR comment - using DatasetId creation to validate referenceName

Use the new class instead of the deprecated one

Revert the unintended changes which will make the test fail

Revert unintended changes

Converting Cassandra source/sinks

Cassandra and Core plugins

Database plugins conversion

Create ReferenceBatchSource/Sink to reuse code

ElasticSearch plugins conversion

HBase plugins conversion

Fix checkstyle

Modifications to HBaseSink and HDFS

Hive Plugins conversion

HTTP, Kafka plugin conversions

MongoDB plugins conversion

Fix referenceName related failures

Hack to use lookup to register a dataset usage

Fix test failures

Use an instance

Remove the old code since they extend Reference* classes

poornachandra commented on core-plugins/widgets/AzureBlobStore-batchsource.json:
```diff
@@ -8,6 +8,11 @@
       "properties": [
         {
           "widget-type": "textbox",
+          "label": "Name for Dataset",
```
What is label? It should not be "Name for Dataset", right? Should it be just "Reference Name"?

gokulavasan replied:
I was not sure what to use for label. Sure, I will use Reference Name.

poornachandra commented:
Like we discussed, we need to make sure referenceName is a valid Id, right?

Other than that code changes look good.

gokulavasan commented:
@awholegunch We need to add the property 'referenceName' to all non-CDAP source/sink plugins. I didn't want to copy paste the description of the property to all .md files. Instead I will leave it to create a doc constant of some kind that we can refer to in all the plugin docs and then modify only that if required.

@poornachandra Thanks for the review Poorna. I am not sure if there is an easy way to make sure that referenceName is a valid CDAP Id. @elfenheart Does UI have such a logic? If yes, we can do that check in UI. Of course, that means the CLI, REST are prone to errors. @sreevatsanraman Any thoughts?

awholegunch commented:
I think the only solution is add this to the Markdown, as there is no pre-processing of these files.

poornachandra commented:
@gokulavasan Even if UI does the id name verification, I think the check should also be done in the plugins. Should we add a helper method in etl-common that can be used to check whether a given string is a valid id? 

@albertshau Any thoughts on how to do this?

gokulavasan commented:
@albertshau @poornachandra The deprecated Id class in cdap-proto has a 'isValidId(String)' method. I can't find an equivalent method in the new EntityId class. If we add one, then I think we can depend on cdap-proto and check if referenceName is a validId during the configurePipeline step and thus fail the creation of the pipeline itself, instead of failing during the MapReduce runtime.

poornachandra commented:
@gokulavasan +1 to the suggestion.

gokulavasan commented:
@poornachandra Can you take a look again? Addressed the comment about validating referenceName property.

poornachandra commented:
LGTM

gokulavasan commented:
Build passed : http://builds.cask.co/browse/HYP-DBT6

albertshau commented:
lgtm

gokulavasan committed:
Lookup operation has to be performed to force the LookupProvider to perform getDataset operation to register lineage

Adding the lookup operation in Sink case as well

Register usage of external dataset using reflection instead of lookup since the latter registers an error log/throws an exception since the dataset is not a KeyValueTable. Though functionally this is alright, it might be misleading to see this error message in the logs

Try/Catch to exclude registering dataset in unit test cases

gokulavasan commented:
@albertshau Please review the last commit. Unfortunately we need to invoke the lookup method on the LookupProvider for it to 'getDataset'. One problem with that is that an exception is thrown (and logged) and the tx is aborted. I will file a JIRA documenting this. This should not be a big overhead since this is done only when the pipeline is initialized. CC @poornachandra

gokulavasan commented:
@albertshau @poornachandra Or we can ignore the lookup that if it is an external dataset in the TxLookupProvider? Is it possible?

albertshau commented:
need to do the same thing for the sink.  Like you mentioned, we may be able to do some hack in the app to avoid the exception. lgtm after adding for sink

gokulavasan commented:
@albertshau @poornachandra Please review the second commit. Modified it to use reflection to invoke getDataset instead of relying on lookupProvider (since using the latter logs an error in the logs which is not ideal and will probably be misleading). Also opened a JIRA CDAP-5759 to fix it.

gokulavasan committed:
Docs for referenceName property

gokulavasan commented:
@awholegunch Please review the third commit. It contains the documentation changes related to the addition of the new referenceName property to external sources and sinks.

awholegunch commented:
@gokulavasan Third commit looks ok to me. I think maybe the order of the properties needs to be addressed, because I think they should bear some resemblance to the UI (and vice-versa) but that can be an issue for another day.

albertshau commented:
lgtm
